lldb/source/Plugins/Platform/WebAssembly/PlatformWasm.cpp

```diff
@@ -0,0 +1,210 @@
+          ArchSpec("wasm64-unknown-unknown-wasm")};
+}
+
+static auto get_arg_range(const Args &args) {
+  return llvm::make_range(args.GetArgumentArrayRef().begin(),
+                          args.GetArgumentArrayRef().end());
+}
+
+lldb::ProcessSP PlatformWasm::DebugProcess(ProcessLaunchInfo &launch_info,
+                                           Debugger &debugger, Target &target,
+                                           Status &error) {
+  Log *log = GetLog(LLDBLog::Platform);
+
+  const PluginProperties &properties = GetGlobalProperties();
+
+  FileSpec runtime = properties.GetRuntimePath();
+  FileSystem::Instance().ResolveExecutableLocation(runtime);
+
+  if (!FileSystem::Instance().Exists(runtime)) {
+    error = Status::FromErrorStringWithFormatv(
+        "WebAssembly runtime does not exist: {0}", runtime.GetPath());
+    return nullptr;
+  }
+
+  uint16_t port = 0;
+  {
+    TCPSocket listen_socket(true);
+    error = listen_socket.Listen("localhost:0", 5);
+    if (error.Fail())
+      return nullptr;
+    port = listen_socket.GetLocalPortNumber();
+  }
+
+  if (error.Fail())
+    return nullptr;
+
+  Args args({runtime.GetPath(),
+             llvm::formatv("{0}{1}", properties.GetPortArg(), port).str()});
+  args.AppendArguments(properties.GetRuntimeArgs());
+  args.AppendArguments(launch_info.GetArguments());
+
+  launch_info.SetArguments(args, true);
+  launch_info.SetLaunchInSeparateProcessGroup(true);
+  launch_info.GetFlags().Clear(eLaunchFlagDebug);
+
+  auto exit_code = std::make_shared<std::optional<int>>();
+  launch_info.SetMonitorProcessCallback(
+      [=](lldb::pid_t pid, int signal, int status) {
+        LLDB_LOG(
+            log,
+            "WebAssembly runtime exited: pid = {0}, signal = {1}, status = {2}",
+            pid, signal, status);
+        exit_code->emplace(status);
+      });
+
+  // This is automatically done for host platform in
+  // Target::FinalizeFileActions, but we're not a host platform.
+  llvm::Error Err = launch_info.SetUpPtyRedirection();
+  LLDB_LOG_ERROR(log, std::move(Err), "SetUpPtyRedirection failed: {0}");
+
+  LLDB_LOG(log, "{0}", get_arg_range(launch_info.GetArguments()));
+  error = Host::LaunchProcess(launch_info);
+  if (error.Fail())
+    return nullptr;
+
+  ProcessSP process_sp = target.CreateProcess(
+      launch_info.GetListener(), wasm::ProcessWasm::GetPluginNameStatic(),
+      nullptr, true);
+  if (!process_sp) {
+    error = Status::FromErrorString("failed to create WebAssembly process");
+    return nullptr;
+  }
+
+  process_sp->HijackProcessEvents(launch_info.GetHijackListener());
+
+  error = process_sp->ConnectRemote(
+      llvm::formatv("connect://localhost:{0}", port).str());
+  if (error.Fail()) {
+    // If we know the runtime has exited, that's a better error message than
+    // failing to connect.
+    if (*exit_code)
+      error = Status::FromErrorStringWithFormatv(
+          "WebAssembly runtime exited with exit code {0}", **exit_code);
```

Comment:
`Status::FromError(llvm::joinErrors(createStringError(...), exit_code.takeError()))`

Reply:
Funny you mention that, it's what I had originally. I ended up going with this because the runtime exiting is always more interesting:

```
(lldb) run
WebAssembly runtime exited with exit code 1
Timed out connecting to localhost:12345
```

But on the other hand it's not nice to drop an error message so I'll revive it.